### PR TITLE
Split supervisor and machine code into separate binaries

### PR DESCRIPTION
**Pull request by fintelia**
This PR breaks off most of the code in the project into a library and then compiles two different executables using it, one for machine mode and one for supervisor mode. The resulting binaries are then linked to one another to produce a single loadable image.

The main trick is to have a constant (SHARED_STATICS) which deferences to a pointer either to physical address 0x80100000 or virtual address 0xffffffffc0100000. The value to use is determined using conditional compilation.

### CODE REVIEW

File: src/machine.rs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    match reason {
+        Some(Reason::EnterSupervisor{ a0, a1, a2, a3, sp, satp, mepc}) => {
+            csrw!(mepc, mepc);
+            csrw!(satp, satp);
+            asm!("mv a0, $0
+                  mv a1, $1
+                  mv a2, $2
+                  mv a3, $3
+                  mv sp, $4
+                  mret" :: "r"(a0), "r"(a1), "r"(a2), "r"(a3), "r"(sp) : "a0", "a1", "a2", "a3", "sp" : "volatile");
```
Comment (celskeggs) [inline src/machine.rs:133]: I'm a little worried about what happens here if rust assigns registers differently than we expect, and these mv's all clobber each other.

File: src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
+
+//! ## Start-up sequence summary:
+//! - QEMU loads hypervisor kernel (this program) and linux kernel (held in initrd) into memory
+//! - QEMU launches hardcoded mrom reset vector, which jumps to 0x80000000
+//! - _start is located at 0x80000000 as the only function in the .init.entrypoint section
+//! - `_start` sets up the stack and calls into mstart
+//! - `mstart` initializes machine-mode control registers as needed by the hypervisor
+//! - `mstart` returns into supervisor-mode in sstart
+//! - `sstart` returns into user-mode at the guest kernel entrypoint
+//!       (running in emulated-supervisor-mode)
```
Comment (celskeggs) [inline src/lib.rs:10]: This comment is now out-of-date; perhaps either update it or remove it?

**Review by celskeggs — APPROVED** [review]
Looks good; here's two things I noticed that you might or might not want to fix.

File: src/machine.rs (hunk 1 above)
Comment (fintelia) [reply src/machine.rs:133]: I finally took the time to report a bug based on that so hopefully it gets fixed upstream: https://github.com/rust-lang/rust/issues/60392.

File: src/lib.rs (hunk 2 above)
Comment (fintelia) [reply src/lib.rs:10]: Good catch

**Pull request merged**